README.md
PALISADE Lattice Cryptography Library
=====================================

PALISADE is a general lattice cryptography library that currently includes efficient implementations of the following lattice cryptography capabilities:
* Fully Homomorphic Encryption (FHE)
   * Brakerski/Fan-Vercauteren (BFV) scheme for integer arithmetic
   * Brakerski-Gentry-Vaikuntanathan (BGV) scheme for integer arithmetic
   * Cheon-Kim-Kim-Song (CKKS) scheme for real-number arithmetic
   * Ducas-Micciancio (FHEW) and Chillotti-Gama-Georgieva-Izabachene (TFHE) schemes for Boolean circuit evaluation
* Multi-Party Extensions of FHE (to support multi-key FHE)
   * Threshold FHE for BGV, BFV, and CKKS schemes
   * Proxy Re-Encryption for BGV, BFV, and CKKS schemes

Note as of version 1.11, the following features have been moved to their own repositories in the PALISADE group.

* Digital Signature (https://gitlab.com/palisade/palisade-signature)
* Identity-Based Encryption (https://gitlab.com/palisade/palisade-abe)
* Ciphertext-Policy Attribute-Based Encryption (https://gitlab.com/palisade/palisade-abe)

All the research prototypes for Key-Policy Attributed-Based Encryption and Program Obfuscation have been moved to https://gitlab.com/palisade/palisade-trapdoor)

PALISADE is a cross-platform C++11 library supporting Linux, Windows, and macOS. The supported compilers are g++ v6.1 or later and clang++ v6.0 or later.

The library also includes unit tests and sample application demos.

PALISADE is available under the BSD 2-clause license.

The library is based on modular architecture with the following layers:

* Math operations layer supporting low-level modular arithmetic, number theoretic transforms, and integer sampling.  This layer is implemented to be portable to multiple hardware computation substrates.
* Lattice operations layer supporting lattice operations, ring algebra, and lattice trapdoor sampling.
* Crypto layer containing efficient implementations of lattice cryptography schemes.
* Encoding layer supporting multiple plaintext encodings for cryptographic schemes.

A major focus is on the usability of the schemes. For instance, all HE schemes with packing use the same common API, and are implemented using runtime polymorphism.

PALISADE implements efficient Residue Number System (RNS) algorithms to achieve high performance, e.g., PALISADE was used as the library for a winning genome-wide association studies solution at iDASH’18.

By default, the library is built without external dependencies. But the user is also provided options to add GMP/NTL, tcmalloc, and/or Intel HEXL third-party libraries if desired.

Further information about PALISADE:

[License Information](LICENSE)

[Library Wiki with documentation](https://gitlab.com/palisade/palisade-development/wikis/home)

[Webinars](Webinars.md)

[YouTube Channel PALISADE](https://www.youtube.com/channel/UC1qByOsQina1rpZ8AGl5TZw)

[Code of Conduct](Code-of-conduct.md)

[Governance](Governance.md)

[Contributing to PALISADE](Contributing.md)

[PALISADE Examples](Examples.md)


Build Instructions
=====================================

We use CMake to build PALISADE. The high-level (platform-independent) procedure for building PALISADE is as follows (for OS-specific instructions, see the section "Detailed information about building PALISADE" at the bottom of this page):

1. Install system prerequisites (if not already installed), including a C++ compiler with OMP support, cmake, make, and autoconf.

2. Clone the PALISADE repo to your local machine.

3. Create a directory where the binaries will be built. The typical choice is a subfolder "build". In this case, the commands are:
```
mkdir build
cd build
cmake ..
```

Note that CMake will check for any system dependencies that are needed for the build process. If the CMake build does not complete successfully, please review the error CMake shows at the end. If the error does not go away (even though you installed the dependency), try running "make clean" to clear the CMake cache.

4. If you want to use any external libraries, such as NTL/GMP or tcmalloc, install these libraries.

5. Build PALISADE by running the following command (this will take few minutes; using the -j make command-line flag is suggested to speed up the build)
```
make
```
If you want to build only library files or some other subset of PALISADE, please review the last paragraph of this page.

After the "make" completes, you should see the PALISADE library files in the lib folder, binaries of demos in bin/demo, binaries of benchmarks in bib/benchmark, and binaries for unit tests in the unittest folder.

6. Install PALISADE to a system directory (if desired or for production purposes)
```
make install
```
You would probably need to run `sudo make install` unless you are specifying some other install location. You can change the install location by running
`cmake -DCMAKE_INSTALL_PREFIX=/your/path ..`. The header files are placed in the "include/palisade" folder of the specified path, and the binaries of the library
are copied directly to the "lib" folder. For example, if no installation path is provided in Ubuntu (and many other Unix-based OSes), the header and library
binary files will be placed in "/usr/local/include/palisade" and "/usr/local/lib", respectively.

Testing and cleaning the build
-------------------

Run unit tests to make sure all capabilities operate as expected
```
make testall
```

Run sample code to test, e.g.,
```
bin/examples/pke/simple-integers
```

To remove the files built by make, you can execute
```
make clean
```

Supported Operating Systems
--------------------------
PALISADE CI continually tests our builds on the following operating systems:

* Ubuntu [18.04] [20.04]
* macOS [Mojave]
* Centos 7
* NVIDIA Xavier [Linux for Tegra 4.2.2]
* MinGW (64-bit) on Windows 10

PALISADE users have reported successful operation on the following systems:

* FreeBSD
* Ubuntu [16.04]
* Arch Linux
* Manjaro Linux

Please let us know the results if you have run PALISADE any additional systems not listed above.

Detailed information about building PALISADE
------------------------------

More detailed steps for some common platforms are provided in the following Wiki articles:

[Instructions for building PALISADE in Linux](https://gitlab.com/palisade/palisade-development/wikis/Instructions-for-building-PALISADE-in-Linux)

[Instructions for building PALISADE in Windows](https://gitlab.com/palisade/palisade-development/wikis/Instructions-for-building-PALISADE-in-Windows)

[Instructions for building PALISADE in macOS](https://gitlab.com/palisade/palisade-development/wikis/Instructions-for-building-PALISADE-in-macOS)

PALISADE provides many CMake/make configuration options, such as installing specific modules of the library, compiling only libraries w/o any unit tests and demos, choosing the Debug mode for compilation, turning on/off NTL/GMP. These options are described in detail in the following Wiki article:

[Use of CMake in PALISADE](https://gitlab.com/palisade/palisade-development/-/wikis/Use-of-CMake-in-PALISADE)

[Instructions for building user projects that use PALISADE](https://gitlab.com/palisade/palisade-development/wikis/Instructions-for-building-user-projects-that-use-PALISADE)


Release_Notes.md
08/22/2021: PALISADE v1.11.4 (stable) is released in the palisade-release repo
* Fixes WebAssembly compilation for new versions of Emscripten (2.0.23+)
* Integrates HEXL v1.2.0 improvements (additional acceleration)
* Adds support for Gaussian sampling with standard deviations between 2^32 and 2^59
* Fixes several minor bugs

05/28/2021: PALISADE v1.11.3 (stable) is released in the palisade-release repo
* Includes all changes from development releases v1.11.0, v1.11.1, and v1.11.2

05/26/2021: PALISADE v1.11.2 (development) is released
* Adds initial Intel HEXL library integration
* Adds support for web assembly
* Fixes some bugs for the scenarios with multiple cryptocontexts

05/04/2021: PALISADE v1.11.1 (development) is released
* Fixes some bugs related to serialization

03/31/2021: PALISADE v1.11.0 (development) is released

* The abe module is moved to https://gitlab.com/palisade/palisade-abe
* The signature module is moved to https://gitlab.com/palisade/palisade-signature
* Adds high-precision CKKS (supports the scaling factor of up to 119 bits in size)
* Removes old code (Stehle-Steinfeld scheme and prior Matrix functionality)
* Applies optimizations for NTT, and hybrid key switching in BGVrns and CKKS
* Simplifies the PALISADE installation (gitmodules are now updated automatically)
* Adds a Dockerfile for PALISADE
* Fixes many bugs reported by the PALISADE community

12/08/2020: PALISADE v1.10.6 (stable) is released

* Patches CKKS against the Li-Micciancio attack
* Complex-number arithmetic is no longer supported in CKKS
* Fixes a bug in the 128-bit mathematical backend
* Fixes a rarely occuring exception in the Gaussian sampling procedure

10/01/2020: PALISADE v1.10.5 (stable) is released

* Fixes a lattice parameter selection bug for an edge case (for leveled HE schemes)
* Fixes a compilation error in Apple clang 12 (XCode 12)

09/21/2020: PALISADE v1.10.4 (stable) is released

* Includes all changes from development releases v1.10.0-v1.10.3
* Fixes some bugs in BGVrns
* Improves error handling

08/25/2020: PALISADE v1.10.3 (development) is released

* Fixes a bug in BFVrns for larger plaintext moduli
* Fixes bugs in FHEW/TFHE
* Fixes bugs affecting the 32-bit and 64-bit native integer backends
* Improves error handling
* Includes minor documentation cleanup

07/11/2020: PALISADE v1.10.2 (development) is released

* Fixes a bug affecting some threshold FHE scenarios
* Adds more validation to CMake flags
* Includes minor documentation cleanup

06/26/2020: PALISADE v1.10.1 (development) is released

* Fixes a bug affecting some clang environments
* Fixes a bug related to the BUILD_BENCHMARKS CMake flag

06/18/2020: PALISADE v1.10.0 (development) is released

* Adds a usable, full RNS variant of BGV (as a more efficient alternative to BFV RNS variants for integer arithmetic)
* Adds a fully functional threshold FHE capability to BGV, BFV, and CKKS to support multi-party computations
* Adds an automated rescaling mode to the approximate rescaling RNS variant of CKKS
* Improves the performance of both CKKS RNS variants
* Improves the performance of both BFV RNS variants
* Makes NTL completely optional for all environments (no quad-precision floats are needed anymore)
* Improves the support for clang in Linux
* Simplifies the CMake procedure for building PALISADE
* Fixes many bugs reported by the PALISADE community

04/24/2020: PALISADE v1.9.2 (stable) is released in the palisade-release repo

* Includes all changes from development releases v1.8.0, v1.9.0, and v1.9.1
* Fixes a bug in BFVrnsB
* Fixes a bug with CSPRNG in some multi-threaded environments

03/03/2020: PALISADE v1.9.1 (development) is released

* Fixes the performance issue with Ubuntu
* Improves the runtime of inverse Number Theoretic Transform for all supported systems

02/29/2020: PALISADE v1.9.0 (development) is released

* Adds multiple low-level optimizations, including improved Number Theoretic Transform
* Adds a CMake install package for PALISADE
* Improves the security and performance of the PRE implementations for BFV, BGV, and CKKS homomorphic encryption schemes
* Updates selected parameter sets for FHEW to reduce the probability of decryption error
* Includes multiple bug fixes for the issues reported by the PALISADE community

01/30/2020: PALISADE v1.8.0 (development) is released

* Adds the Gama-Izabachene-Nguyen-Xie (GINX) bootstrapping to the FHEW implementation
* Includes other FHEW optimizations: now the bootstrapping key size is 20x times smaller and runtime about 2.5x faster
* Adds XOR and XNOR gates to FHEW

01/22/2020: PALISADE v1.7.4 (stable) is released in the palisade-release repo

* includes all changes from development releases v1.7.a to v1.7.d

01/15/2020: PALISADE v1.7.d (development) is released

* Replaces the Mersenne Twister PRNG Engine with a cryptographically secure BLAKE2-based PRNG
* Fixes a PRNG bug affecting environments where multithreading is done not using OpenMP

12/26/2019: PALISADE v1.7.c (development) is released

* Fixes a bug affecting applications using the PALISADE serialization capability
* Changes the installation paths for header files and library binary files

12/19/2019: PALISADE v1.7.b (development) is released

* Simplifies the development of applications using PALISADE (only one header file is now needed for most applications)
* Changes the folder hierarchy of the library header files

11/15/2019: PALISADE v1.7.a (development) is released

* Adds an optimized implementation of the CKKS scheme for approximate (real-number) homomorphic encryption. The implementation features automated rescaling/modswitching and hybrid key switching.
* Adds an optimized implementation of the FHEW scheme for Boolean-circuit homomorphic encryption, supporting standard HE parameters. Performs bootstrapping for each binary gate, and hence supports the evaluation of arbitrary Boolean circuits.
* Adds the ring dimension as an option when generating crypto contexts for BFV variants.

src/core/extras/dft.cpp
// @file dft.cpp - Example of Discrete Fourier Transform
// @author TPOC: contact@palisade-crypto.org
//
// @copyright Copyright (c) 2019, New Jersey Institute of Technology (NJIT)
// All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
// 1. Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution. THIS SOFTWARE IS
// PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
// EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include "math/dftransfrm.h"

using namespace lbcrypto;

int main() {
  std::vector<std::complex<double>> dftVec(64);
  dftVec.at(0) = std::complex<double>(4, 0);
  dftVec.at(1) = std::complex<double>(5, 0);
  dftVec.at(2) = std::complex<double>(5, 0);
  dftVec.at(3) = std::complex<double>(4.2, 0);
  dftVec.at(4) = std::complex<double>(5, 0);
  dftVec.at(5) = std::complex<double>(7.1, 0);
  dftVec.at(6) = std::complex<double>(6, 0);
  dftVec.at(7) = std::complex<double>(3, 0);
  dftVec.at(8) = std::complex<double>(4, 0);
  dftVec.at(9) = std::complex<double>(5, 0);
  dftVec.at(10) = std::complex<double>(5, 0);
  dftVec.at(11) = std::complex<double>(4.2, 0);
  dftVec.at(12) = std::complex<double>(5, 0);
  dftVec.at(13) = std::complex<double>(7.1, 0);
  dftVec.at(14) = std::complex<double>(6, 0);
  dftVec.at(15) = std::complex<double>(3, 0);
  dftVec.at(16) = std::complex<double>(4, 0);
  dftVec.at(17) = std::complex<double>(5, 0);
  dftVec.at(18) = std::complex<double>(5, 0);
  dftVec.at(19) = std::complex<double>(4.2, 0);
  dftVec.at(20) = std::complex<double>(5, 0);
  dftVec.at(21) = std::complex<double>(7.1, 0);
  dftVec.at(22) = std::complex<double>(6, 0);
  dftVec.at(23) = std::complex<double>(3, 0);
  dftVec.at(24) = std::complex<double>(4, 0);
  dftVec.at(25) = std::complex<double>(5, 0);
  dftVec.at(26) = std::complex<double>(5, 0);
  dftVec.at(27) = std::complex<double>(4.2, 0);
  dftVec.at(28) = std::complex<double>(5, 0);
  dftVec.at(29) = std::complex<double>(7.1, 0);
  dftVec.at(30) = std::complex<double>(6, 0);
  dftVec.at(31) = std::complex<double>(3, 0);
  dftVec.at(32) = std::complex<double>(4, 0);
  dftVec.at(33) = std::complex<double>(5, 0);
  dftVec.at(34) = std::complex<double>(5, 0);
  dftVec.at(35) = std::complex<double>(4.2, 0);
  dftVec.at(36) = std::complex<double>(5, 0);
  dftVec.at(37) = std::complex<double>(7.1, 0);
  dftVec.at(38) = std::complex<double>(6, 0);
  dftVec.at(39) = std::complex<double>(3, 0);
  dftVec.at(40) = std::complex<double>(4, 0);
  dftVec.at(41) = std::complex<double>(5, 0);
  dftVec.at(42) = std::complex<double>(5, 0);
  dftVec.at(43) = std::complex<double>(4.2, 0);
  dftVec.at(44) = std::complex<double>(5, 0);
  dftVec.at(45) = std::complex<double>(7.1, 0);
  dftVec.at(46) = std::complex<double>(6, 0);
  dftVec.at(47) = std::complex<double>(3, 0);
  dftVec.at(48) = std::complex<double>(4, 0);
  dftVec.at(49) = std::complex<double>(5, 0);
  dftVec.at(50) = std::complex<double>(5, 0);
  dftVec.at(51) = std::complex<double>(4.2, 0);
  dftVec.at(52) = std::complex<double>(5, 0);
  dftVec.at(53) = std::complex<double>(7.1, 0);
  dftVec.at(54) = std::complex<double>(6, 0);
  dftVec.at(55) = std::complex<double>(3, 0);
  dftVec.at(56) = std::complex<double>(4, 0);
  dftVec.at(57) = std::complex<double>(5, 0);
  dftVec.at(58) = std::complex<double>(5, 0);
  dftVec.at(59) = std::complex<double>(4.2, 0);
  dftVec.at(60) = std::complex<double>(5, 0);
  dftVec.at(61) = std::complex<double>(7.1, 0);
  dftVec.at(62) = std::complex<double>(6, 0);
  dftVec.at(63) = std::complex<double>(3, 0);

  DiscreteFourierTransform::PreComputeTable(128);

  double start = currentDateTime();
  std::vector<std::complex<double>> dftVec2 =
      DiscreteFourierTransform::ForwardTransform(dftVec);
  double end = currentDateTime();
  std::cout << "Without table: " << end - start << " ms" << std::endl;

  start = currentDateTime();
  std::vector<std::complex<double>> dftVec3 =
      DiscreteFourierTransform::ForwardTransform(dftVec);
  end = currentDateTime();
  std::cout << "With table: " << end - start << " ms" << std::endl << std::endl;
}
